hkdf.go

```diff
@@ -125,69 +110,176 @@ func (c *hkdf) Read(p []byte) (int, error) {
-		return nil, newOpenSSLError("EVP_PKEY_derive")
-	}
-	return out[:r.keylen], nil
 }
 
 func ExpandHKDF(h func() hash.Hash, pseudorandomKey, info []byte) (io.Reader, error) {
-	c, err := newHKDF(h, C.GO_EVP_KDF_HKDF_MODE_EXPAND_ONLY)
+	if !SupportsHKDF() {
+		return nil, errUnsupportedVersion()
+	}
+
+	md, err := hashFuncToMD(h)
 	if err != nil {
 		return nil, err
 	}
+
 	switch vMajor {
-	case 3:
-		if C.go_openssl_EVP_PKEY_CTX_set1_hkdf_key(c.ctx,
-			base(pseudorandomKey), C.int(len(pseudorandomKey))) != 1 {
-			return nil, newOpenSSLError("EVP_PKEY_CTX_set1_hkdf_key")
-		}
-		if C.go_openssl_EVP_PKEY_CTX_add1_hkdf_info(c.ctx,
-			base(info), C.int(len(info))) != 1 {
-			return nil, newOpenSSLError("EVP_PKEY_CTX_add1_hkdf_info")
-		}
 	case 1:
-		if C.go_openssl_EVP_PKEY_CTX_ctrl(c.ctx, -1, C.GO1_EVP_PKEY_OP_DERIVE,
-			C.GO_EVP_PKEY_CTRL_HKDF_KEY,
-			C.int(len(pseudorandomKey)), unsafe.Pointer(base(pseudorandomKey))) != 1 {
-			return nil, newOpenSSLError("EVP_PKEY_CTX_set1_hkdf_key")
+		ctx, err := newHKDFCtx1(md, C.GO_EVP_KDF_HKDF_MODE_EXPAND_ONLY, nil, nil, pseudorandomKey, info)
+		if err != nil {
+			return nil, err
 		}
-		if C.go_openssl_EVP_PKEY_CTX_ctrl(c.ctx, -1, C.GO1_EVP_PKEY_OP_DERIVE,
-			C.GO_EVP_PKEY_CTRL_HKDF_INFO,
-			C.int(len(info)), unsafe.Pointer(base(info))) != 1 {
-			return nil, newOpenSSLError("EVP_PKEY_CTX_add1_hkdf_info")
+		c := &hkdf1{ctx: ctx, hashLen: int(C.go_openssl_EVP_MD_get_size(md))}
+		runtime.SetFinalizer(c, (*hkdf1).finalize)
+		return c, nil
+	case 3:
+		ctx, err := newHKDFCtx3(md, C.GO_EVP_KDF_HKDF_MODE_EXPAND_ONLY, nil, nil, pseudorandomKey, info)
+		if err != nil {
+			return nil, err
 		}
+		c := &hkdf3{ctx: ctx, hashLen: int(C.go_openssl_EVP_MD_get_size(md))}
+		runtime.SetFinalizer(c, (*hkdf3).finalize)
+		return c, nil
+	default:
+		panic(errUnsupportedVersion())
 	}
-	return c, nil
+}
+
+type hkdf3 struct {
+	ctx C.GO_EVP_KDF_CTX_PTR
+
+	hashLen int
+	buf     []byte
+}
+
+func (c *hkdf3) finalize() {
+	if c.ctx != nil {
+		C.go_openssl_EVP_KDF_CTX_free(c.ctx)
+	}
+}
+
+// fetchHKDF3 fetches the HKDF algorithm.
+// It is safe to call this function concurrently.
+// The returned EVP_KDF_PTR shouldn't be freed.
+var fetchHKDF3 = sync.OnceValues(func() (C.GO_EVP_KDF_PTR, error) {
+	checkMajorVersion(3)
+
+	name := C.CString("HKDF")
+	kdf := C.go_openssl_EVP_KDF_fetch(nil, name, nil)
+	C.free(unsafe.Pointer(name))
+	if kdf == nil {
+		return nil, newOpenSSLError("EVP_KDF_fetch")
+	}
+	return kdf, nil
+})
+
+// newHKDFCtx3 implements HKDF for OpenSSL 3 using the EVP_KDF API.
+func newHKDFCtx3(md C.GO_EVP_MD_PTR, mode C.int, secret, salt, pseudorandomKey, info []byte) (C.GO_EVP_KDF_CTX_PTR, error) {
+	checkMajorVersion(3)
+
+	kdf, err := fetchHKDF3()
+	if err != nil {
+		return nil, err
+	}
+	ctx := C.go_openssl_EVP_KDF_CTX_new(kdf)
+	if ctx == nil {
+		return nil, newOpenSSLError("EVP_KDF_CTX_new")
+	}
+	defer func() {
+		if err != nil {
```

Comment:
Miss making `err` a named return?

Seems like the other return changes are going in this direction for `ctx`. (Although I prefer `(_ C.GO_EVP_KDF_CTX_PTR, err error)`, like `newCipherCtx`, so you don't need to worry about it.)

(No bug currently, just easier to introduce one later by shadowing+returning `err` in a deeper scope.)

Reply:
Fixed, thanks!